File: desktopclient/src/main/java/com/goexp/galgame/gui/view/common/StateChangeController.java
package com.goexp.galgame.gui.view.common;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.view.task.ChangeGameTask;
import com.goexp.galgame.gui.model.Game;
import javafx.beans.value.ChangeListener;
import javafx.collections.FXCollections;
import javafx.concurrent.Service;
import javafx.concurrent.Task;
import javafx.fxml.FXML;
import javafx.scene.control.ChoiceBox;
import javafx.util.StringConverter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Comparator;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class StateChangeController {

    private final Logger logger = LoggerFactory.getLogger(StateChangeController.class);

    @FXML
    private ChoiceBox<GameState> choiceState;

    private ChangeListener<GameState> listener;
    private Game targetGame;
    private Service<Void> changeGameStateService = new Service<>() {
        @Override
        protected Task<Void> createTask() {
            return new ChangeGameTask.Like(targetGame);
        }
    };

    @FXML
    private void initialize() {
        var types = Stream.of(
                GameState.values()
        )
                .sorted(Comparator.reverseOrder())
                .collect(Collectors.toList());

        choiceState.setConverter(new StringConverter<>() {
            public String toString(GameState object) {
                return object.getName();
            }

            public GameState fromString(String string) {
                return GameState.from(string);
            }
        });

        choiceState.setItems(FXCollections.observableArrayList(types));


        listener = (o, oldValue, newValue) -> {

            if (newValue != null) {
                logger.debug("<Action>Value:{},New:{}", choiceState.getValue(), newValue);

                targetGame.state.set(newValue);

                changeGameStateService.restart();
            }
        };
    }

    public void load(Game game) {
        choiceState.valueProperty().removeListener(listener);
        choiceState.setValue(game.state.get());
        targetGame = game;
        choiceState.valueProperty().addListener(listener);
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/part/StateChangeController.scala
package com.goexp.galgame.gui.view.game.part

import com.goexp.galgame.common.model.game.{GameLocation, GameState}
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.task.game.change.{Block, State}
import com.goexp.ui.javafx.{DefaultController, TaskService}
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.ChoiceBox
import scalafx.Includes._

import scala.jdk.CollectionConverters._

class StateChangeController extends DefaultController {
  @FXML private var choiceState: ChoiceBox[GameState] = _

  var targetGame: Game = _

  final lazy val state = new SimpleObjectProperty[GameState]


  final private val changeGameStateService = TaskService(new State(targetGame))
  final private val blockService = TaskService(new Block(targetGame))

  override protected def initComponent() = {
    val types = GameState.values.to(LazyList).sortBy(gs => gs.value)(Ordering[Int].reverse).asJava
    choiceState.setItems(FXCollections.observableArrayList(types))
  }


  override protected def dataBinding(): Unit = {
    choiceState.value <==> state
  }


  override protected def eventBinding(): Unit = {

    state.onChange((_, oldValue, newValue) => {
      if (oldValue != null && newValue != null) {
        logger.debug(s"<Action>Value:${choiceState.getValue},New:${newValue}")

        //reset some info
        if (newValue eq GameState.BLOCK) {
          targetGame.star.set(0)
          targetGame.location.set(GameLocation.REMOTE)

          blockService.restart()
        } else {
          changeGameStateService.restart()
        }

      }
    })
  }
}